src/Type/PHPUnit/MockObjectDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,42 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\PHPUnit;
+
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\Type\Generic\GenericObjectType;
+use PHPStan\Type\IntersectionType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+use PHPUnit\Framework\MockObject\InvocationMocker;
+use PHPUnit\Framework\MockObject\MockObject;
+
+class MockObjectDynamicReturnTypeExtension implements \PHPStan\Type\DynamicMethodReturnTypeExtension
+{
+
+	public function getClass(): string
+	{
+		return MockObject::class;
+	}
+
+	public function isMethodSupported(MethodReflection $methodReflection): bool
+	{
+		return $methodReflection->getName() === 'expects';
+	}
+
+	public function getTypeFromMethodCall(MethodReflection $methodReflection, MethodCall $methodCall, Scope $scope): Type
+	{
+		$type = $scope->getType($methodCall->var);
+		if (!($type instanceof IntersectionType)) {
+			return new GenericObjectType(InvocationMocker::class, []);
+		}
+
+		$mockClass = array_filter($type->getTypes(), function (Type $type): bool {
+			return !($type instanceof ObjectType && $type->getClassName() === MockObject::class);
+		});
```

Comment:
Make sure the array isn't empty; please name the variable as a plural.

Reply:
Can you change the condition into:

```php
!$type instanceof TypeWithClassName || $type->getClassName() !== MockObject::class
```

Reply:
> Make sure the array isn't empty

why ?

I'll return 
```
return new GenericObjectType(InvocationMocker::class, []);
```

This sounds ok to me

Reply:
That's invalid.